libraries/AP_Scripting/generator/src/main.c

```diff
@@ -1506,60 +1571,21 @@ void emit_userdata_method(const struct userdata *data, const struct method *meth
     fprintf(source, "    AP::scheduler().get_semaphore().give();\n");
   }
 
-  int return_count = 1; // number of arguments to return
+  // we need to emit out refernce arguments, iterate the args again, creating and copying objects, while keeping a new count
+  int return_count = 1; 
+  if (method->flags & TYPE_FLAGS_REFERNCE) {
+    arg = method->arguments;
+    // number of arguments to return
+    return_count += emit_references(arg,"    ");
+  }
+
   switch (method->return_type.type) {
     case TYPE_BOOLEAN:
       if (method->flags & TYPE_FLAGS_NULLABLE) {
         fprintf(source, "    if (data) {\n");
         // we need to emit out nullable arguments, iterate the args again, creating and copying objects, while keeping a new count
-        return_count = 0;
         arg = method->arguments;
-        int arg_index = NULLABLE_ARG_COUNT_BASE + 2;
-        while (arg != NULL) {
-          if (arg->type.flags & TYPE_FLAGS_NULLABLE) {
-            return_count++;
-            switch (arg->type.type) {
-              case TYPE_BOOLEAN:
-                fprintf(source, "        lua_pushboolean(L, data_%d);\n", arg_index);
-                break;
-              case TYPE_FLOAT:
-                fprintf(source, "        lua_pushnumber(L, data_%d);\n", arg_index);
-                break;
-              case TYPE_INT8_T:
-              case TYPE_INT16_T:
-              case TYPE_INT32_T:
-              case TYPE_UINT8_T:
-              case TYPE_UINT16_T:
-              case TYPE_ENUM:
-                fprintf(source, "        lua_pushinteger(L, data_%d);\n", arg_index);
-                break;
-              case TYPE_UINT32_T:
-                fprintf(source, "        new_uint32_t(L);\n");
-                fprintf(source, "        *static_cast<uint32_t *>(luaL_checkudata(L, -1, \"uint32_t\")) = data_%d;\n", arg_index);
-                break;
-              case TYPE_STRING:
-                fprintf(source, "        lua_pushstring(L, data_%d);\n", arg_index);
-                break;
-              case TYPE_USERDATA:
-                // userdatas must allocate a new container to return
-                fprintf(source, "        new_%s(L);\n", arg->type.data.ud.sanatized_name);
-                fprintf(source, "        *check_%s(L, -1) = data_%d;\n", arg->type.data.ud.sanatized_name, arg_index);
-                break;
-              case TYPE_NONE:
-                error(ERROR_INTERNAL, "Attempted to emit a nullable argument of type none");
-                break;
-              case TYPE_LITERAL:
-                error(ERROR_INTERNAL, "Attempted to make a nullable literal");
-                break;
-              case TYPE_AP_OBJECT: // FIXME: collapse these to a single failure case
-                error(ERROR_INTERNAL, "Attempted to make a nullable ap_object");
-                break;
-            }
-          }
-
-          arg_index++;
-          arg = arg->next;
-        }
+        return_count = 0 + emit_references(arg,"        ");
```

Comment:
```suggestion
        return_count = emit_references(arg,"        ");
```

Reply:
done, not sure how that happend in the first place...